### PR TITLE
HipMRI Study Using VQVAE - 47323254

### PR DESCRIPTION
**Pull request by ReonaUQ**
This pull request introduces the VQVAE model designed for the HipMRI image study. The model aims to transform the input image into the latent space, quantise it, and reconstruct it to learn discrete representation of images effectively. I have experimented with various settings including learning rates, data augmentations, model complexities, and downscaling factors. The quality of the model performance was evaluated with Structural Similarity Index Measure (SSIM) score and this project achieved SSIM of over 0.9 with the downscaling of 4 times in 35 minutes of training.

It includes scripts for training, validation, and evaluation of the model, as well as a README file that documents the project overview including instructions for executing the scripts, and conducted experiments. The project is organised into three main directories: `src`, which contains all the scripts needed to train, validate, and evaluate the model; `configs`, which contains sample configuration YAML files for training and prediction; and `images`, which contains all the images used in the README.

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Very clear and highly resembling reconstructed images, with an SSIM well above 0.9 (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

4. Design:
Not bad design and good practice using utils script (1 Mark)

5. Commenting:
Sufficient docstrings but comments could be more specific (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- It's really nice to have progressive experiments and hyperparameter tuning recorded with consequent detailed results :) (it would be better and easier to see the resulting plots if they load properly in README.md);
- Images not shown properly on README.md due to incorrect path configuration;
- It is ok to train and predict the way you did in `train.py` and `predict.py`, but it's better to have specific functions to do so for modularization purpose;
- No need to delete recognition/README.md
- Good project!

**Comment by YashMitttal** [comment]
Additional suggestions for improvements:

- While you've mentioned the reconstruction and commitment loss, explaining why these are used and how they contribute to model performance would be good.


Overall very good job.

**Comment by ReonaUQ** [comment]
Would it be possible to modify the implementation slightly? I'd like to add a commitment loss term (| z - sg(e) |^2) since, currently, I only have the reconstruction and codebook loss. I initially didn't included it because the GitHub code I referenced didn’t include it, and I didn’t encounter stability issues during training. However, upon further consideration, there is no reason to not include it as it ensures the training stability without much extra computation.

**Comment by hanemma7moud** [comment]
Reona Nakagawa

**Comment by shakes76** [comment]
Good work. Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.
